src/rt/sched/backpressure.h

```diff
@@ -0,0 +1,259 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+#pragma once
+
+/**
+ * ## Introduction
+ *
+ * The primary goal of backpressure is to prevent runaway growth of a cown's
+ * message queue that might eventually result in a Verona application running
+ * out of memory. This is achieved by tracking the load on each cown to detect
+ * when it is unable to process messages as quickly as messages are added to the
+ * queue. Once a cown is considered "overloaded", message sends to that cown
+ * will result in the senders being muted so that the overloaded cown may
+ * temporarily process more messages than it receives. Similarly, cowns sending
+ * to muted receivers will also be muted so that cowns don't experience runaway
+ * queue growth while they are muted. Muted cowns are generally not rescheduled
+ * until the receiver that resulted in their muting (the "mutor") is no longer
+ * muted or overloaded.
+ *
+ * ## Cown Load
+ *
+ * The backpressure system uses "load" as an approximation for message queue
+ * depth. Load is measured using a token message that occasionally falls out of
+ * its message queue. The "current load" on a cown is equal to the number of
+ * messages processed by the cown since the last token message fell out. Once
+ * the next token falls out of its queue, the current load is stored into a ring
+ * buffer and the current load is reset. The total load for the cown is
+ * calculated as the sum of the current load and previous loads stored in the
+ * ring buffer. Simply measuring the batch size of messages processed by the
+ * cown does not account for additional participants of a message cutting their
+ * batch short once they receive the mutli-message.
+ *
+ * ## Mute Map Scan
+ *
+ * Once a scheduler thread completes a message action that would result in the
+ * muted multimessage cowns, the cowns are acquired by the scheduler thread and
+ * placed in a "mute map" on the scheduler thread. The mute map is a mapping
+ * from mutor => mute set, where "mute set" referrs to the set of cowns muted by
+ * a mutor. Each scheduler thread scans the mutors of its mute map on each
+ * iteration of the run loop. If any mutor is not muted and is not overloaded,
+ * the corresponding mute set is unmuted and the entry is removed from the mute
+ * map. A mutor may exist as the key in mutiple mute maps, but a muted cown may
+ * only be tracked in a single mute set.
+ *
+ * ## Message Scan
+ *
+ * All messages from one set of sending cowns to a set of receiving cowns is
+ * scanned to determine if the senders should be muted. Such a message will
+ * result in muted senders if all of the following are true:
+ *   1. The receiver set does not contain any of the senders
+ *   2. No sender is overloaded
+ *   3. One of the receivers is either overloaded or muted.
+ *
+ * A message is given "priority" if either condition 1 or 2 are false. A
```

Comment:
wait, what? 

this sounds wrong. can you explain this?

Reply:
so a sender isn't overloaded. the receiver is muted.

so the receiver will be unmuted?

Reply:
No, in this case conditions 2 and 3 are true. So no muting occurs.

Reply:
Ok, I think the way it is phrased is very confusing i finally figured out how it is supposed to read. That was really really confusing to me. Too much negation I think.

Reply:
I think saying "a sender is overloaded" would be a lot more clear (and then not having the negation).

Reply:
Going over yet again, I think that you should write out the rules for muting in one section and the rules for priority on another section. I think the mixing them together results in some interesting negations and double negations that are really hard to parse.